test/fixtures/es-module-loaders/assertionless-json-import.mjs

```diff
@@ -0,0 +1,18 @@
+const DATA_URL_PATTERN = /^data:application\/json(?:[^,]*?)(;base64)?,([\s\S]*)$/;
+const JSON_URL_PATTERN = /\.json(\?[^#]*)?(#.*)?$/;
+
+export function resolve(specifier, context, defaultResolve) {
+  if (DATA_URL_PATTERN.test(specifier)) {
+    const { importAssertions } = context;
+    importAssertions.type = 'json';
```

Comment:
> Does it mean that if the object is mutated (without being returned), then the mutated version is sent to the `load` hook? This is what I am concerned about.

I guess it is?

https://github.com/nodejs/node/blob/2c61806e07d028777ed11aaa336a8242bc3354d9/lib/internal/modules/esm/loader.js#L236-L242

The `importAssertions` declared here is what’s later passed into the `load` hook.

So a user `resolve` hook receives a reference to `originalImportAssertions`. If the user hook mutates that and doesn’t return it, then `importAssertions = resolvedImportAssertions || originalImportAssertions` would be interpreted as `importAssertions = undefined || theMutatedOriginalImportAssertions`. Which, yeah, is I guess what you’re concerned about. This wasn’t intentional, it was more that I didn’t consider the case you’re describing.

An easy enough solution to this would be to clone `originalImportAssertions` before passing it into the `resolve` hook. Then if the user hook mutates it, those edits are lost unless the user hook returns it. I’m fine with that if you want.

But please explain a little more detail why the “user mutates the object” concern is problematic?

Reply:
> I think I prefer the explicit return. I also think that the documentation should be clear about what the hook is supposed to do with this object.

The explicit return makes the API more complicated that it needs to be, so I went ahead and get rid of it.
What the hook is suppose to do:
- `resolve()` should not do anything with it, according to https://github.com/nodejs/loaders/issues/39#issuecomment-935400756. We're still leaving it here in case some use case shows up (but could be removed before the loaders are unflagged).
- `load()` should use it to determine if a module can load or not. I've added a line in the documentation.

If you check my example of assertion-less JSON import loader, you can see how it can be useful to mutate the import assertion object. Note that it would also be possible for this hook to re-create an object before passing it to the next hook, but that seems unnecessary.
If you have a reason for the object to be frozen, we can still change that (loaders are still flagged, so we can even change this after this PR lands if something ever comes up).

Reply:
Again, what I'm concerned about isn't really that the object is mutable. It's whether mutating it can have side-effect. I don't want the behavior of the loaders to depend on side-effects.

Reply:
Passing a copy of the original object to the hook sgtm.

Reply:
Fair enough, take a look at https://github.com/nodejs/node/pull/40250/commits/e082f7f79cbe2dffdab3840a10361864ba3efa98.